compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -602,4 +612,56 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         }
         self.flounder(&responses)
     }
+
+    #[instrument(level = "debug", skip(self), ret)]
+    pub(super) fn evaluate_alias_bound_self_is_well_formed<G: GoalKind<'tcx>>(
```

Comment:
This is the logic we had discussed before, essentially that an alias bound only holds if the trait ref of the alias is satisfied via a param-env candidate.